File: backend_python/controllers/userProgram.py
from flask import Blueprint, request, jsonify
from models.userProgram import UserProgram

up_bp = Blueprint('up', __name__)
up_model = None

@up_bp.route('/get', methods=['GET'])
def get_up():
    try:
        up = up_model.get_all_up()
        print("Up in get_all_up route:", up)
        return jsonify(success=True, data=up), 200
    except Exception as e:
        return jsonify(success=False, message=str(e)), 500

@up_bp.route('/add', methods=['POST'])
def add_up():
    data = request.get_json()
    try:
        up_model.add_up(data['up_uid'], data['up_name'], data['up_description'], data['duration'], data['time'])
        return jsonify(success=True, message="Up added successfully"), 200
    except Exception as e:
        return jsonify(success=False, message=str(e)), 500

@up_bp.route('/updateName/<id>', methods=['PUT'])
def update_name(id):
    data = request.get_json()
    try:
        up_model.update_name(id, data['up_name'])
        return jsonify(success=True, message="Name updated successfully"), 200
    except Exception as e:
        return jsonify(success=False, message=str(e)), 500

@up_bp.route('/updateDescription/<id>', methods=['PUT'])
def update_description(id):
    data = request.get_json()
    try:
        up_model.update_description(id, data['up_description']), 200
        return jsonify(success=True, message="Description updated successfully"), 200
    except Exception as e:
        return jsonify(success=False, message=str(e)), 500

@up_bp.route('/updateDuration/<id>', methods=['PUT'])
def update_duration(id):
    data = request.get_json()
    try:
        up_model.update_duration(id, data['duration']), 200
        return jsonify(success=True, message="Duration updated successfully"), 200
    except Exception as e:
        return jsonify(success=False, message=str(e)), 500

@up_bp.route('/updateTime/<id>', methods=['PUT'])
def update_time(id):
    data = request.get_json()
    try:
        up_model.update_time(id, data['time']), 200
        return jsonify(success=True, message="Time updated successfully"), 200
    except Exception as e:
        return jsonify(success=False, message=str(e)), 500

@up_bp.route('/delete/<id>', methods=['PATCH'])
def delete_up(id):
    try:
        up_model.delete_up(id)
        return jsonify(success=True, message="Up deleted successfully"), 200
    except Exception as e:
        return jsonify(success=False, message=str(e)), 500

def initialize_userProgram_model(model):
    global up_model
    up_model = model

File: client/src/services/userProgram.js
import axios from 'axios';
const UP = "http://localhost:3000/userProgram";

const get = async (id) => {
    try{
        const response = await axios.get(`${UP}/get/${id}`);
        return response.data;
    } catch (error){
        console.error('Error fetching user program:', error);
        throw error;
    }
};

const add = async (upData) => {
    try{
        const response = await axios.post(`${UP}/add`, upData);
        return response.data;
    } catch (error){
        console.error('Error adding user program:', error);
        throw error;
    }
};

const updateName = async (id, name) => {
    try{
        const response = await axios.put(`${UP}/updateName/${id}`, name);
        return response.data;
    } catch (error){
        console.error('Error updating user program:', error);
        throw error;
    }
};

const updateDescription = async (id, desc) => {
    try{
        const response = await axios.put(`${UP}/updateDescription/${id}`, desc);
        return response.data;
    } catch (error){
        console.error('Error updating user program:', error);
        throw error;
    }
};

const updateDuration = async (id, duration) => {
    try{
        const response = await axios.put(`${UP}/updateDuration/${id}`, duration);
        return response.data;
    } catch (error){
        console.error('Error updating user program:', error);
        throw error;
    }
};

const updateTime = async (id, time) => {
    try{
        const response = await axios.put(`${UP}/updateTime/${id}`, time);
        return response.data;
    } catch (error){
        console.error('Error updating user program:', error);
        throw error;
    }
};

const deleted = async (id) => {
    try{
        const response = await axios.patch(`${UP}/delete/${id}`);
        return response.data;
    } catch (error){
        console.error('Error deleting user program:', error);
        throw error;
    }
};

const byId = async (id) => {
    try{
        const response = await axios.get(`${UP}/byId/${id}`);
        return response.data;
    } catch (error){
        console.error('Error fetching user program by id:', error);
        throw error;
    }
};

const total = async (id) => {
    try{
        const response = await axios.get(`${UP}/total/${id}`);
        return response.data;
    } catch (error){
        console.error('Error fetching total:', error);
        throw error;
    }
}

const totalAll = async () => {
    try{
        const response = await axios.get(`${UP}/totalAll`);
        return response.data;
    } catch (error){
        console.error('Error fetching total:', error);
        throw error;
    }
}

const hwServices = {
    get,
    add,
    updateName,
    updateDescription,
    updateDuration,
    updateTime,
    deleted,
    byId, 
    total,
    totalAll
};

export default hwServices;